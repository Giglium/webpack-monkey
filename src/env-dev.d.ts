/// <reference types="webpack/module" />
/// <reference types="webpack-dev-server" />

declare const __PORT__: number
