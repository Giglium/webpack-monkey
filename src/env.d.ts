declare namespace webpack {
  interface Hot {
    monkeyReload(options?: import("./client/hmr").MonkeyReloadOptions): void
  }
}
