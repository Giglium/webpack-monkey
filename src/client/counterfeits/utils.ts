export function parseHeaders(headers: string): Record<string, string> {
  return headers
    .split("\n")
    .filter(Boolean)
    .map((line) => line.split(": "))
    .reduce((headers, [name, value]) => {
      headers[name!] = value!
      return headers
    }, {} as Record<string, string>)
}
