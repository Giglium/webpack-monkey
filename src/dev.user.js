;(async function () {
  "use strict"

  const { clientScript, runtimeScript } = __MK_DEV_INJECTION__

  if (!clientScript || !runtimeScript) {
    console.warn("[Monkey Dev] client script not found")
    return
  }

  window.__MK_GLOBAL__ = unsafeWindow.__MK_GLOBAL__ = {
    loadScript: (url) => {
      return new Promise((resolve, reject) => {
        const fail = () => {
          reject(
            new Error(`failed to load script (${url}), please check if the server is running.`)
          )
        }

        GM_xmlhttpRequest({
          method: "GET",
          url,
          onload: (res) => {
            try {
              if (res.status === 404) {
                throw new Error(`Script not found (${url}).`)
              }

              const fakeScript = document.createElement("script")
              fakeScript.src = url

              Object.defineProperty(document, "currentScript", {
                get: () => fakeScript,
                configurable: true,
              })

              eval(res.responseText)

              delete document.currentScript

              resolve()
            } catch (e) {
              reject(e)
            }
          },
          onerror: fail,
          ontimeout: fail,
        })
      })
    },
  }

  console.log("[Monkey Dev] running")

  __MK_GLOBAL__.loadScript(runtimeScript).then(() => __MK_GLOBAL__.loadScript(clientScript))
})()
