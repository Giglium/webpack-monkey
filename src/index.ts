declare global {
  namespace webpack {
    interface Hot {
      monkeyReload(options?: import("./client/hmr").MonkeyReloadOptions): void
    }
  }
}

export * from "./node/monkey"
