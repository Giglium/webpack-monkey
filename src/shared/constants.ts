export const VAR_MK_INJECTION = "__MK_INJECTION__"
export const VAR_MK_DEV_INJECTION = "__MK_DEV_INJECTION__"

export const DEV_SCRIPT = "monkey-dev.user.js"
export const CLIENT_SCRIPT = "monkey-client.js"
