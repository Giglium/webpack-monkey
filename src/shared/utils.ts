import { matchPattern } from "browser-extension-url-match"
import { isNil } from "lodash"

/**
 * array.includes() with type guard.
 */
export function includes<T>(array: readonly T[], value: any): value is T {
  return array.includes(value)
}

export function urlMatch(pattern: string, url: string) {
  const matcher = matchPattern(pattern)

  if (!matcher.valid) {
    throw new TypeError("Invalid pattern: " + pattern)
  }

  return matcher.match(url)
}

export function parentUntil<T>(
  target: T | null | undefined,
  getParent: (target: T) => T | null | undefined,
  callback: (target: T) => void | boolean
): T | null | undefined {
  const visited = new Set<T>()

  let current = target
  let result = null

  while (!isNil(current) && !visited.has(current)) {
    visited.add(current)

    if (callback(current)) {
      result = current
      return result
    }

    current = getParent(current)
  }

  return result
}
