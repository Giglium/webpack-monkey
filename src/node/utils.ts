// @ts-ignore
// prettier-ignore
import { getDescriptionFile, getRequiredVersionFromDescriptionFile } from "webpack/lib/sharing/utils";

import { isArray, isBoolean, isNil, isObject, isString, trimEnd } from "lodash"
import { Compilation, sources } from "webpack"
import { getGMAPIs } from "../shared/GM"
import { META_FIELDS, META_FIELDS_I18N, UserscriptMeta } from "../shared/meta"
import { includes } from "../shared/utils"

export function getPackageDepVersion(packageJson: any, dep: string): string | undefined {
  if (!isObject(packageJson)) {
    return undefined
  }

  return getRequiredVersionFromDescriptionFile(packageJson, dep)
}

export function getPackageJson(fs: Compilation["inputFileSystem"], context: string) {
  return new Promise<{ data: object; path: string }>((resolve, reject) => {
    getDescriptionFile(fs, context, ["package.json"], (err: any, result: any) => {
      if (err) {
        reject(err)
      } else {
        resolve(result)
      }
    })
  })
}

export function pathSplit(path: string) {
  return path.replace(/\\/g, "/").split("/")
}

export function traverseAndFindSource(
  source: unknown,
  cb: (source: sources.Source) => void | true,
): sources.Source | undefined {
  if (!source || !(source instanceof sources.Source)) {
    return undefined
  }

  const found = cb(source)

  if (found === true) {
    return source
  }

  if ((source as any)._children) {
    for (const child of (source as any)._children as unknown[]) {
      const result = traverseAndFindSource(child, cb)

      if (result) {
        return result
      }
    }
  }

  return traverseAndFindSource((source as any)._source, cb)
}

export function generateMetaBlock(source: string, meta: UserscriptMeta) {
  let metaBlock = "// ==UserScript==\n"
  const fieldPrefix = "// @"

  const metaFields = Object.keys(meta)
  const maxFieldLength = Math.max(
    ...["grant", "require", ...metaFields].map((field) => {
      if (includes(META_FIELDS_I18N, field)) {
        const value = meta[field]

        if (isObject(value)) {
          const maxLangLength = Math.max(
            ...Object.keys(value).map((lang) => (lang === "default" ? 0 : lang.length)),
          )

          return field.length + maxLangLength
        }
      }

      return field.length
    }),
  )
  const indentSize = fieldPrefix.length + maxFieldLength + 2
  const indentEnd = (str: string) => str.padEnd(indentSize, " ")

  function putField(field: string, value: string | string[] | boolean) {
    let line = fieldPrefix + field

    if (isString(value)) {
      line = indentEnd(line) + value
    } else if (isArray(value)) {
      if (!value.length) {
        return
      }

      line = value.map((v) => indentEnd(line) + v).join("\n")
    } else if (isBoolean(value)) {
      // ignore false value
      if (!value) {
        return
      }
    } else {
      console.warn("Unknown type of value:", value)
    }

    metaBlock += line + "\n"
  }

  for (const field of META_FIELDS) {
    if (includes(META_FIELDS_I18N, field)) {
      const value = meta[field]

      if (value) {
        if (isString(value)) {
          putField(field, value)
        } else {
          for (const lang of Object.keys(value)) {
            putField(field + (lang === "default" ? "" : ":" + lang), value[lang]!)
          }
        }
      }
    } else if (field === "grant") {
      putField(
        "grant",
        getGMAPIs().filter((api) => source.includes(api)),
      )

      if (meta.grant) {
        putField("grant", meta.grant)
      }
    } else {
      if (!isNil(meta[field])) {
        putField(field, meta[field]!)
      }
    }
  }

  metaBlock += "// ==/UserScript=="

  return metaBlock
}

export interface ResolveExternalResult {
  type?: string
  identifier?: string
  url: string
  value: string
}

export function resolveUrlExternal(externalValue: string): ResolveExternalResult | undefined {
  // full example: "var foo@https://example.com"
  const match = externalValue.match(/^(.*? )?(.+?@)?(https?:\/\/.+)$/)

  if (!match) {
    return undefined
  }

  const type = match[1]?.trim()
  const identifier = match[2] && trimEnd(match[2], "@")
  const url = match[3]!
  const value = (type ? type + " " : "") + (identifier || JSON.stringify(url))

  const resolved: ResolveExternalResult = {
    type,
    identifier,
    url,
    value,
  }

  return resolved
}

export function getUnnamedUrlExternalErrorMessage(url?: string) {
  const exampleUrl = url || "https://example.com"
  return `Unexpected reference to unnamed external module with URL "${url || "<unknown URL>"}".
This happens when you import a module from a URL
but do not specify an identifier for it,
e.g. \`import foo from "${exampleUrl}"\`, which will cause
runtime errors in the generated code. To fix this,
either add a universally unique identifier to the import statement,
e.g. \`import foo from "foo@${exampleUrl}"\`,
or do not import anything from it,
e.g. \`import "${exampleUrl}"\`.`.replace(/\n/g, " ")
}
