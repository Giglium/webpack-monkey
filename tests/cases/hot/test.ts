import path from "path"
import { Page } from "puppeteer"
import Server from "webpack-dev-server"
import { merge } from "webpack-merge"
import { webpackMonkey } from "../../../src"
import { DEV_SCRIPT } from "../../../src/shared/constants"
import { installWithTampermonkey } from "../../utils/tampermonkey"
import {
  testBuild,
  usingDevServerHot,
  withCommonConfig,
  withMiniCssExtract,
} from "../../utils/webpack"

const entry = path.resolve(__dirname, "index.js")
const depA = path.resolve(__dirname, "depA.js")
const depB = path.resolve(__dirname, "depB.js")
const depC = path.resolve(__dirname, "depC.js")
const depCss = path.resolve(__dirname, "styles.css")

const config = withCommonConfig({
  mode: "development",
  entry: entry,
  output: {
    path: path.resolve(__dirname, "dist"),
  },
  devServer: {
    port: __PORT__,
  },
})

it("build", () => testBuild(webpackMonkey(withMiniCssExtract(config))))

describe("hot reload", () => {
  it.browser("modules", async () => {
    await usingDevServerHot(
      { config: webpackMonkey(withMiniCssExtract(config)) },
      async (server, { replacers }) => {
        await installWithTampermonkey(browser, page, `http://localhost:${__PORT__}/${DEV_SCRIPT}`)
        await page.goto(`http://localhost:${__PORT__}/webpack-dev-server/`)

        const expectExactlyOnes = async (classes: string[]) => {
          const selector = classes.map((c) => `.${c}`).join(",")
          const received = await page.$$eval(selector, (els) => els.map((el) => el.className))
          expect(received).toIncludeSameMembers(classes)
        }

        {
          await page.waitForSelector(".index1")

          replacers[entry] = (s) => s.replace("index1", "index2")
          await invalidate(server)
          await page.waitForSelector(".index2")

          expect(await page.$(".index1")).toBeNull()

          delete replacers[entry]
        }
        {
          await expectExactlyOnes(["depA1", "depB1", "depC1"])

          replacers[depA] = (s) => s.replace("depA1", "depA2")
          await invalidate(server)
          await waitForInsertion(page, ".index1")

          await expectExactlyOnes(["depA2", "depB1", "depC1"])

          replacers[depC] = (s) => s.replace("depC1", "depC2")
          await invalidate(server)
          await waitForInsertion(page, ".index1")

          await expectExactlyOnes(["depA2", "depB1", "depC2"])
        }
      }
    )
  })

  it.browser.each([
    ["css with mini-css-extract", withMiniCssExtract()],
    // ["css with style-loader", withStyleLoader()],
  ])("%s", async (_, cssConfig) => {
    await usingDevServerHot(
      { config: webpackMonkey(merge({}, config, cssConfig)) },
      async (server, { replacers }) => {
        await installWithTampermonkey(browser, page, `http://localhost:${__PORT__}/${DEV_SCRIPT}`)
        await page.goto(`http://localhost:${__PORT__}/webpack-dev-server/`)
        await page.waitForSelector(".index1")

        const bodyBgColor = await page.evaluate(
          () => getComputedStyle(document.body).backgroundColor
        )
        expect(bodyBgColor).toBe("rgb(255, 0, 0)")

        replacers[depCss] = (s) =>
          s.replace("background-color: red;", "").replace("color: white;", "color: black;")
        await invalidate(server)

        await expect(
          page.waitForFunction(() => {
            const styles = getComputedStyle(document.body)
            return styles.backgroundColor === "rgba(0, 0, 0, 0)" && styles.color === "rgb(0, 0, 0)"
          })
        ).resolves.toBeTruthy()
      }
    )
  })
})

function invalidate(server: Server) {
  return new Promise((r) => server.invalidate(r))
}

function waitForInsertion(page: Page, selector: string, observingSelector = "body") {
  return page.waitForFunction(
    (selector, observingSelector) => {
      return new Promise<HTMLElement>((r) => {
        const observer = new MutationObserver((mutations) => {
          mutations.forEach((mutation) => {
            mutation.addedNodes.forEach((node) => {
              if (node instanceof HTMLElement && node.matches(selector)) {
                observer.disconnect()
                r(node)
              }
            })
          })
        })
        observer.observe(document.querySelector(observingSelector)!, { childList: true })
      })
    },
    {},
    selector,
    observingSelector
  )
}
