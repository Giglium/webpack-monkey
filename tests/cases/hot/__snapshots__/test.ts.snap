// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build 1`] = `
"// ==UserScript==
// @name     Hot test
// @grant    GM_addStyle
// @grant    GM_addElement
// @match    *://*/*
// @version  1.0.0
// ==/UserScript==

;(() => {
  "use strict"

  GM_addElement(document.body, "div", { class: "depA1" })

  GM_addElement(document.body, "div", { class: "depC1" })

  GM_addElement(document.body, "div", { class: "depB1" })

  GM_addElement(document.body, "div", { class: "index1" })
})()

GM_addStyle(\`
body {
  background-color: red;
  color: white;
}

\`)
"
`;
